jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java

```diff
@@ -0,0 +1,419 @@
+                                }
+                            }
+                        }
+                    }
+
+            }
+            // update k and the corresponding sets
+            k += 1;
+            pathsLengthK = pathsLengthKplus1;
+            pathsLengthKplus1 = new LinkedHashMap<>();
+        }
+
+        return new GraphWalk<>(graph, new ArrayList<>(lengthBound), Double.MAX_VALUE);
+    }
+
+    /**
+     * Checks whether <code>path</code> dominates the current minimal cost path with the same head, tail and label set
+     * in the set of all paths of length k + 1. Thus, dominated paths are eliminated.
+     * This is important out of efficiency reasons, otherwise many unnecessary paths may
+     * be considered in further calculations.
+     *
+     * @param path the currently calculated path
+     * @param pathsLengthKplus1 all before calculated paths of length k + 1
+     *
+     * @return whether <code>path</code> dominates the current minimal cost path with the same head, tail and label set.
+     */
+    private boolean removeDominatedPathsOfLengthKplus1(LabeledPath<V> path, Map<PathSetKey<V>, LabeledPath<V>> pathsLengthKplus1) {
+        // simulates domination test by using the index structure
+        LabeledPath<V> pathToCheck = pathsLengthKplus1.get(PathSetKey.of(path.getHead(), path.getTail(), path.getLabels()));
+        if (pathToCheck != null) {
+            return pathToCheck.getCost() < path.getCost();
+        }
+        return false;
+    }
+
+    /**
+     * Checks whether <code>path</code> is dominated by some path in the before calculated set of paths of length k.
+     * This is important out of efficiency reasons, otherwise many unnecessary paths may
+     * be considered in further calculations.
+     *
+     * @param path the currently calculated path
+     * @param pathsLengthK all before calculated paths of length k
+     *
+     * @return whether <code>path</code> is dominated by some path in <code>pathsLengthK</code>
+     */
+    private boolean removeDominatedPathsOfLengthK(LabeledPath<V> path, Map<PathSetKey<V>, LabeledPath<V>> pathsLengthK) {
+        Set<Integer> modifiableLabelSet = new HashSet<>(path.getLabels());
+        for(Integer label : path.getLabels()) {
+            modifiableLabelSet.remove(label);
+            // simulates domination test by using the index structure
+            LabeledPath<V> pathToCheck = pathsLengthK.get(PathSetKey.of(path.getHead(), path.getTail(), modifiableLabelSet));
+            if (pathToCheck != null) {
+                if(pathToCheck.getCost() < path.getCost()) {
+                    return true;
+                }
+            }
+            modifiableLabelSet.add(label);
+        }
+        return false;
+    }
+
+    /**
+     * Adds a path to the data structure <code>paths</code>, which contains all paths indexed by their head and tail, such that they are sorted by their cost.
```

Comment:
I suggest renaming this method to "updatePathIndex" since now it does an add-or-replace rather than just an add.